controllers/propagator/replicatedpolicy_controller.go

```diff
@@ -0,0 +1,445 @@
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+	appsv1 "open-cluster-management.io/multicloud-operators-subscription/pkg/apis/apps/placementrule/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	policiesv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
+	"open-cluster-management.io/governance-policy-propagator/controllers/common"
+)
+
+var _ reconcile.Reconciler = &ReplicatedPolicyReconciler{}
+
+type ReplicatedPolicyReconciler struct {
+	Propagator
+}
+
+func (r *ReplicatedPolicyReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	log := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	log.Info("Reconciling the replicated policy")
+
+	// Set the hub template watch metric after reconcile
+	defer func() {
+		hubTempWatches := r.DynamicWatcher.GetWatchCount()
+		log.V(3).Info("Setting hub template watch metric", "value", hubTempWatches)
+
+		hubTemplateActiveWatchesMetric.Set(float64(hubTempWatches))
+	}()
+
+	replicatedExists := true
+	replicatedPolicy := &policiesv1.Policy{}
+
+	if err := r.Get(ctx, request.NamespacedName, replicatedPolicy); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			log.Error(err, "Failed to get the replicated policy")
+
+			return reconcile.Result{}, err
+		}
+
+		replicatedExists = false
+	}
+
+	rootName, rootNS, err := common.ParseRootPolicyLabel(request.Name)
+	if err != nil && replicatedExists {
+		if err := r.cleanUpReplicated(ctx, replicatedPolicy); err != nil {
+			if !k8serrors.IsNotFound(err) {
+				log.Error(err, "Failed to delete the invalid replicated policy, requeueing")
+
+				return reconcile.Result{}, err
+			}
+		}
+
+		log.Info("Invalid replicated policy deleted")
+
+		return reconcile.Result{}, nil
+	}
+
+	// Fetch the Root Policy instance
+	rootPolicy := &policiesv1.Policy{}
+	rootNN := types.NamespacedName{Namespace: rootNS, Name: rootName}
+
+	if err := r.Get(ctx, rootNN, rootPolicy); err != nil {
+		if k8serrors.IsNotFound(err) {
+			if replicatedExists {
+				// do not handle a replicated policy which does not belong to the current cluster
+				inClusterNS, err := common.IsInClusterNamespace(r.Client, request.Namespace)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+
+				if !inClusterNS {
+					log.Info("Found a replicated policy in non-cluster namespace, skipping it")
+
+					return reconcile.Result{}, nil
+				}
+
+				// otherwise, we need to clean it up
+				if err := r.cleanUpReplicated(ctx, replicatedPolicy); err != nil {
+					if !k8serrors.IsNotFound(err) {
+						log.Error(err, "Failed to delete the orphaned replicated policy, requeueing")
+
+						return reconcile.Result{}, err
+					}
+				}
+
+				log.Info("Orphaned replicated policy deleted")
+
+				return reconcile.Result{}, nil
+			}
+
+			log.Info("Root policy and replicated policy already missing")
```

Comment:
This may be a better "debug" log.

Reply:
I'll be looking through all the log calls here for https://issues.redhat.com/browse/ACM-7778... I will make this (and the others you mention) one level higher, but if we notice more, let's just hope they covered by that work.